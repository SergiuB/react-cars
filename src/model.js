import PropTypes from 'prop-types';

const model = {
  name: {
    type: 'string',
  },
  acceleration: {
    type: 'number',
    min: 1,
    max: 100,
    step: 0.1,
  },
  cylinders: {
    type: 'number',
    min: 2,
    max: 16,
    step: 1,
  },
  displacement: {
    type: 'number',
    min: 1,
    max: 1000,
    step: 1,
  },
  horsepower: {
    type: 'number',
    min: 40,
    max: 1000,
    step: 1,
  },
  mpg: {
    type: 'number',
    min: 0,
    max: 100,
    step: 1,
  },
  weight: {
    type: 'number',
    min: 500,
    max: 10000,
    step: 1,
  },
  year: {
    type: 'number',
    min: 0,
    max: 99,
    step: 1,
  },
};


const carShape = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  acceleration: PropTypes.number,
  cylinders: PropTypes.number,
  displacement: PropTypes.number,
  horsepower: PropTypes.number,
  mpg: PropTypes.number,
  weight: PropTypes.number,
  year: PropTypes.number,
});

const fieldShape = PropTypes.shape({
  type: PropTypes.string.isRequired,
  min: PropTypes.number,
  max: PropTypes.number,
  step: PropTypes.number,
});

const modelShape = PropTypes.shape({
  name: fieldShape,
  acceleration: fieldShape,
  cylinders: fieldShape,
  displacement: fieldShape,
  horsepower: fieldShape,
  mpg: fieldShape,
  weight: fieldShape,
  year: fieldShape,
});

export {
  carShape,
  modelShape,
};

export default model;
