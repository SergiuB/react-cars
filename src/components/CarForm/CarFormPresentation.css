.car-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  border: 1px;
}

.car-form-fields {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  margin: 10px;
}

.car-form-buttons button {
  margin: 5px;
}