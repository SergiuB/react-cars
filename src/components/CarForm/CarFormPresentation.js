import React from 'react';
import PropTypes from 'prop-types';
import './CarFormPresentation.scss';

const CarFormPresentation = ({
    name,
    nameError,
    onChangeName,
    acceleration,
    accelerationError,
    onChangeAcceleration,
    onSubmit,
    onCancel,
  }) => {
    const hasError = !!(nameError || accelerationError);
    return (
      <form onSubmit={onSubmit}>
        <input type="text" name="name" onChange={onChangeName} value={name}/>
          { nameError && <span className="error">{ nameError }</span> }
        <input type="number" name="acceleration" onChange={onChangeAcceleration} value={acceleration}/>
          { accelerationError && <span className="error">{ accelerationError }</span> }
        <button type="submit" name="submit" disabled={hasError}>Save</button>
        <button type="button" name="cancel" onClick={onCancel}>Cancel</button>
      </form>
    );
  }

CarFormPresentation.propTypes = {
  name: PropTypes.string.isRequired,
  nameError: PropTypes.string.isRequired,
  acceleration: PropTypes.number.isRequired,
  accelerationError: PropTypes.string.isRequired,
  onChangeName: PropTypes.func.isRequired,
  onChangeAcceleration: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

CarFormPresentation.defaultProps = {
  nameError: '',
  accelerationError: '',
}

export default CarFormPresentation;