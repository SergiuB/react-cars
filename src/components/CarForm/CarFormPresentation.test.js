import React from 'react';
import { shallow } from 'enzyme';

import CarFormPresentation from './CarFormPresentation';

import model from '../../model';

describe('CarFormPresentation', function () {
  it('renders correctly', function () {
    const onChangeField = jest.fn();
    const onSubmit = jest.fn();
    const onCancel = jest.fn();
    const wrapper = shallow(
      <CarFormPresentation
        car={{
          name: 'Audi',
          acceleration: 12,
        }}
        model={model}
        onChangeField={onChangeField}
        onSubmit={onSubmit}
        onCancel={onCancel}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with field errors but cannot submit', function () {
    const onChangeField = jest.fn();
    const onSubmit = jest.fn();
    const onCancel = jest.fn();
    const wrapper = shallow(
      <CarFormPresentation
        car={{
          name: 'Audi',
          acceleration: 12,
        }}
        errors={{
          name: 'Not a good name',
          acceleration: 'This car sucks',
        }}
        model={model}
        onChangeField={onChangeField}
        onSubmit={onSubmit}
        onCancel={onCancel}
      />
    );
    expect(wrapper).toMatchSnapshot();

    wrapper.find('[type="submit"]').simulate('click');

    // submit button should be disabled in this case
    expect(onSubmit).not.toBeCalled();
  });


  it('fires the handlers passed in props', function () {
    const onChangeField = jest.fn();
    const onSubmit = jest.fn();
    const onCancel = jest.fn();
    const wrapper = shallow(
      <CarFormPresentation
        car={{
          name: 'Audi',
          acceleration: 12,
        }}
        model={model}
        onChangeField={onChangeField}
        onSubmit={onSubmit}
        onCancel={onCancel}
      />
    );

    wrapper.find('input[name="acceleration"]').simulate('change', { target: { value: 3 } });
    expect(onChangeField).toBeCalledWith('acceleration', 3);

    wrapper.find('input[name="name"]').simulate('change', { target: { value: 'VW' } });
    expect(onChangeField).toBeCalledWith('name', 'VW');

    wrapper.find('button[name="cancel"]').simulate('click');
    expect(onCancel).toBeCalled();

    wrapper.find('form').simulate('submit');
    expect(onSubmit).toBeCalled();
  });
});
