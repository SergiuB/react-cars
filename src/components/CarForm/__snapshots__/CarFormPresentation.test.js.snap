// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarFormPresentation renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <form
    className="car-form border border-primary"
    onSubmit={[Function]}
>
    <div
        className="car-form-fields"
    >
        <label>
            Name 
            <input
                name="name"
                onChange={[Function]}
                type="text"
                value="Audi"
            />
        </label>
        
        <label>
            Acceleration 
            <input
                name="acceleration"
                onChange={[Function]}
                step="0.1"
                type="number"
                value={12}
            />
        </label>
        
    </div>
    <div
        className="car-form-buttons"
    >
        <button
            className="text-primary"
            disabled={false}
            name="submit"
            type="submit"
        >
            Save
        </button>
        <button
            name="cancel"
            onClick={[Function]}
            type="button"
        >
            Cancel
        </button>
    </div>
</form>,
  "nodes": Array [
    <form
      className="car-form border border-primary"
      onSubmit={[Function]}
>
      <div
            className="car-form-fields"
      >
            <label>
                  Name 
                  <input
                        name="name"
                        onChange={[Function]}
                        type="text"
                        value="Audi"
                  />
            </label>
            
            <label>
                  Acceleration 
                  <input
                        name="acceleration"
                        onChange={[Function]}
                        step="0.1"
                        type="number"
                        value={12}
                  />
            </label>
            
      </div>
      <div
            className="car-form-buttons"
      >
            <button
                  className="text-primary"
                  disabled={false}
                  name="submit"
                  type="submit"
            >
                  Save
            </button>
            <button
                  name="cancel"
                  onClick={[Function]}
                  type="button"
            >
                  Cancel
            </button>
      </div>
</form>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <CarFormPresentation
        car={
                Object {
                        "acceleration": 12,
                        "name": "Audi",
                      }
        }
        changeHandlers={
                Object {
                        "acceleration": [Function],
                        "name": [Function],
                      }
        }
        errors={
                Object {
                        "acceleration": "",
                        "name": "",
                      }
        }
        onCancel={[Function]}
        onSubmit={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "car": Object {
            "acceleration": 12,
            "name": "Audi",
          },
          "changeHandlers": Object {
            "acceleration": [Function],
            "name": [Function],
          },
          "errors": Object {
            "acceleration": "",
            "name": "",
          },
          "onCancel": [Function],
          "onSubmit": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <form
          className="car-form border border-primary"
          onSubmit={[Function]}
>
          <div
                    className="car-form-fields"
          >
                    <label>
                              Name 
                              <input
                                        name="name"
                                        onChange={[Function]}
                                        type="text"
                                        value="Audi"
                              />
                    </label>
                    
                    <label>
                              Acceleration 
                              <input
                                        name="acceleration"
                                        onChange={[Function]}
                                        step="0.1"
                                        type="number"
                                        value={12}
                              />
                    </label>
                    
          </div>
          <div
                    className="car-form-buttons"
          >
                    <button
                              className="text-primary"
                              disabled={false}
                              name="submit"
                              type="submit"
                    >
                              Save
                    </button>
                    <button
                              name="cancel"
                              onClick={[Function]}
                              type="button"
                    >
                              Cancel
                    </button>
          </div>
</form>,
        "_debugID": 2,
        "_renderedOutput": <form
          className="car-form border border-primary"
          onSubmit={[Function]}
>
          <div
                    className="car-form-fields"
          >
                    <label>
                              Name 
                              <input
                                        name="name"
                                        onChange={[Function]}
                                        type="text"
                                        value="Audi"
                              />
                    </label>
                    
                    <label>
                              Acceleration 
                              <input
                                        name="acceleration"
                                        onChange={[Function]}
                                        step="0.1"
                                        type="number"
                                        value={12}
                              />
                    </label>
                    
          </div>
          <div
                    className="car-form-buttons"
          >
                    <button
                              className="text-primary"
                              disabled={false}
                              name="submit"
                              type="submit"
                    >
                              Save
                    </button>
                    <button
                              name="cancel"
                              onClick={[Function]}
                              type="button"
                    >
                              Cancel
                    </button>
          </div>
</form>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CarFormPresentation
    car={
        Object {
            "acceleration": 12,
            "name": "Audi",
          }
    }
    changeHandlers={
        Object {
            "acceleration": [Function],
            "name": [Function],
          }
    }
    errors={
        Object {
            "acceleration": "",
            "name": "",
          }
    }
    onCancel={[Function]}
    onSubmit={[Function]}
/>,
}
`;

exports[`CarFormPresentation renders correctly with field errors but cannot submit 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <form
    className="car-form border border-primary"
    onSubmit={[Function]}
>
    <div
        className="car-form-fields"
    >
        <label>
            Name 
            <input
                name="name"
                onChange={[Function]}
                type="text"
                value="Audi"
            />
        </label>
        <span
            className="text-danger"
        >
            <small>
                Not a good name
            </small>
        </span>
        <label>
            Acceleration 
            <input
                name="acceleration"
                onChange={[Function]}
                step="0.1"
                type="number"
                value={12}
            />
        </label>
        <span
            className="text-danger"
        >
            <small>
                This car sucks
            </small>
        </span>
    </div>
    <div
        className="car-form-buttons"
    >
        <button
            className="text-primary"
            disabled={true}
            name="submit"
            type="submit"
        >
            Save
        </button>
        <button
            name="cancel"
            onClick={[Function]}
            type="button"
        >
            Cancel
        </button>
    </div>
</form>,
  "nodes": Array [
    <form
      className="car-form border border-primary"
      onSubmit={[Function]}
>
      <div
            className="car-form-fields"
      >
            <label>
                  Name 
                  <input
                        name="name"
                        onChange={[Function]}
                        type="text"
                        value="Audi"
                  />
            </label>
            <span
                  className="text-danger"
            >
                  <small>
                        Not a good name
                  </small>
            </span>
            <label>
                  Acceleration 
                  <input
                        name="acceleration"
                        onChange={[Function]}
                        step="0.1"
                        type="number"
                        value={12}
                  />
            </label>
            <span
                  className="text-danger"
            >
                  <small>
                        This car sucks
                  </small>
            </span>
      </div>
      <div
            className="car-form-buttons"
      >
            <button
                  className="text-primary"
                  disabled={true}
                  name="submit"
                  type="submit"
            >
                  Save
            </button>
            <button
                  name="cancel"
                  onClick={[Function]}
                  type="button"
            >
                  Cancel
            </button>
      </div>
</form>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <CarFormPresentation
        car={
                Object {
                        "acceleration": 12,
                        "name": "Audi",
                      }
        }
        changeHandlers={
                Object {
                        "acceleration": [Function],
                        "name": [Function],
                      }
        }
        errors={
                Object {
                        "acceleration": "This car sucks",
                        "name": "Not a good name",
                      }
        }
        onCancel={[Function]}
        onSubmit={[Function]}
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "car": Object {
            "acceleration": 12,
            "name": "Audi",
          },
          "changeHandlers": Object {
            "acceleration": [Function],
            "name": [Function],
          },
          "errors": Object {
            "acceleration": "This car sucks",
            "name": "Not a good name",
          },
          "onCancel": [Function],
          "onSubmit": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <form
          className="car-form border border-primary"
          onSubmit={[Function]}
>
          <div
                    className="car-form-fields"
          >
                    <label>
                              Name 
                              <input
                                        name="name"
                                        onChange={[Function]}
                                        type="text"
                                        value="Audi"
                              />
                    </label>
                    <span
                              className="text-danger"
                    >
                              <small>
                                        Not a good name
                              </small>
                    </span>
                    <label>
                              Acceleration 
                              <input
                                        name="acceleration"
                                        onChange={[Function]}
                                        step="0.1"
                                        type="number"
                                        value={12}
                              />
                    </label>
                    <span
                              className="text-danger"
                    >
                              <small>
                                        This car sucks
                              </small>
                    </span>
          </div>
          <div
                    className="car-form-buttons"
          >
                    <button
                              className="text-primary"
                              disabled={true}
                              name="submit"
                              type="submit"
                    >
                              Save
                    </button>
                    <button
                              name="cancel"
                              onClick={[Function]}
                              type="button"
                    >
                              Cancel
                    </button>
          </div>
</form>,
        "_debugID": 4,
        "_renderedOutput": <form
          className="car-form border border-primary"
          onSubmit={[Function]}
>
          <div
                    className="car-form-fields"
          >
                    <label>
                              Name 
                              <input
                                        name="name"
                                        onChange={[Function]}
                                        type="text"
                                        value="Audi"
                              />
                    </label>
                    <span
                              className="text-danger"
                    >
                              <small>
                                        Not a good name
                              </small>
                    </span>
                    <label>
                              Acceleration 
                              <input
                                        name="acceleration"
                                        onChange={[Function]}
                                        step="0.1"
                                        type="number"
                                        value={12}
                              />
                    </label>
                    <span
                              className="text-danger"
                    >
                              <small>
                                        This car sucks
                              </small>
                    </span>
          </div>
          <div
                    className="car-form-buttons"
          >
                    <button
                              className="text-primary"
                              disabled={true}
                              name="submit"
                              type="submit"
                    >
                              Save
                    </button>
                    <button
                              name="cancel"
                              onClick={[Function]}
                              type="button"
                    >
                              Cancel
                    </button>
          </div>
</form>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CarFormPresentation
    car={
        Object {
            "acceleration": 12,
            "name": "Audi",
          }
    }
    changeHandlers={
        Object {
            "acceleration": [Function],
            "name": [Function],
          }
    }
    errors={
        Object {
            "acceleration": "This car sucks",
            "name": "Not a good name",
          }
    }
    onCancel={[Function]}
    onSubmit={[Function]}
/>,
}
`;
